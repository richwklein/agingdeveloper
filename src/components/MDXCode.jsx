import React from "react";
import {Highlight, themes} from "prism-react-renderer";
import {ChildrenProps} from "../props";

/**
 * React component used by the {@link MDXProvider} when rendering code in mdx.
 *
 * @param {MDXCodeProps} props - The code tag props.
 * @return {React.ReactElement} - The react component
 */
export const MDXCode = ({children}) => {
  if (children.type != "code") {
    return (
      <pre>{children}</pre>
    );
  }

  const {props: {className, children: code}} = children;
  const language = className?.replace(/language-/, "").trim() || "";

  return (
    <Highlight
      theme={themes.vsDark}
      code={code.trim()}
      language={language}
    >
      {({className, style, tokens, getLineProps, getTokenProps}) => (
        <pre className={className} style={{...style}}>
          {tokens.map((line, index) => {
            const lineProps = getLineProps({line, key: index});
            return (
              <div key={index} {...lineProps}>
                {line.map((token, key) => (
                  <span key={key}{...getTokenProps({token, key})} />
                ))}
              </div>
            );
          },
          )}

        </pre>
      )}
    </Highlight>
  );
};

/**
 * @typedef MDXCodeProps - The code tag props
 * @property {ChildrenProps} children - The children of the tag.
 */
MDXCode.propTypes = {
  children: ChildrenProps,
};

export default MDXCode;
