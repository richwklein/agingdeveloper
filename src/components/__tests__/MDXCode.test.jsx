import React from "react";
import {cleanup, render} from "@testing-library/react";
import MDXCode from "../MDXCode";


describe("MDXCode", () => {
  afterEach(cleanup);

  const renderComponent = (children) => render(
      <MDXCode>
        {children}
      </MDXCode>,
  );

  test("that it renders correctly.", () => {
    const codeBlock = React.createElement(
        "code",
        {"className": "language-js"},
        "Module not found: Error: Can't resolve '@mdx-jsreact'\n");

    const {container} = renderComponent(codeBlock);
    expect(container.firstChild).toMatchSnapshot();
  });

  test("that a non-codeblock is wrapped in a pre tag.", () => {
    const div = React.createElement("div", {}, "This is a div block");
    const {container} = renderComponent(div);
    expect(container.firstChild).toContainHTML("<pre><div>This is a div block</div></pre>");
  });

  test("that a string is wrapped in a pre tag.", () => {
    const {container} = renderComponent("content");
    expect(container.firstChild).toContainHTML("<pre>content</pre>");
  });
});
