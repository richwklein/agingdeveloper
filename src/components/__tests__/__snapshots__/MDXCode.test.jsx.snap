// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MDXCode that it renders correctly. 1`] = `
<pre
  class="prism-code language-js"
  style="color: rgb(156, 220, 254); background-color: rgb(30, 30, 30);"
>
  <div
    class="token-line"
    style="color: rgb(156, 220, 254);"
  >
    <span
      class="token maybe-class-name"
    >
      Module
    </span>
    <span
      class="token plain"
    >
       not found
    </span>
    <span
      class="token operator"
      style="color: rgb(212, 212, 212);"
    >
      :
    </span>
    <span
      class="token plain"
    >
       
    </span>
    <span
      class="token known-class-name class-name"
      style="color: rgb(78, 201, 176);"
    >
      Error
    </span>
    <span
      class="token operator"
      style="color: rgb(212, 212, 212);"
    >
      :
    </span>
    <span
      class="token plain"
    >
       
    </span>
    <span
      class="token maybe-class-name"
    >
      Can
    </span>
    <span
      class="token string"
      style="color: rgb(206, 145, 120);"
    >
      't resolve '
    </span>
    <span
      class="token plain"
    >
      @mdx
    </span>
    <span
      class="token operator"
      style="color: rgb(212, 212, 212);"
    >
      -
    </span>
    <span
      class="token plain"
    >
      jsreact'
    </span>
  </div>
</pre>
`;
